doc/rst/technotes/optimization.rst

```diff
@@ -0,0 +1,662 @@
+  most improvement to the program's performance. You can print out the
+  performance of each region after it is timed. It's a good idea to use a
+  ``config const`` to make the timing only print when you are studying
+  performance.
+
+Counting Communication Events
+  If you need to do most of your work by running locally oversubscribed
+  (see above) then it's particularly useful to measure performance by
+  counting the number of communication events. Counting communication
+  events can also be useful as a secondary performance metric for
+  multi-node runs. Reducing the number of communication events is a useful
+  strategy for optimizing communication.
+
+  You can use the :chpl:mod:`CommDiagnostics` module and
+  ``startCommDiagnostics()`` followed by ``getCommDiagnostics()`` to
+  count communication events. It's common to
+  ``writeln(getCommDiagnostics()``.  Note that the communication count
+  information will be easier to understand if you compile with
+  ``--no-cache-remote``.
+
+  Communication counts provide a way to compare communication
+  performance independent of where you are running. For example, you
+  might measure and seek to reduce the communication counts when running
+  oversubscribed on a laptop or workstation. Reductions in communication
+  counts from such an effort should also help with multi-node performance
+  when you are able to run on a big system.
+
+
+Overall Flow
+~~~~~~~~~~~~
+
+The general flow of performance optimization will be:
+
+ * use techniques like timing the different phases of your computation to
+   try to understand which parts are the slow or not getting faster as you
+   add more locales or cores
+ * come up with a theory as to what could be improved by thinking about
+   what is happening when your program is running and/or by looking at
+   timing or communication count information
+ * make an adjustment (the adjustment only needs to be good enough to
+   measure its impact on performance at this step)
+ * measure its impact on performance
+ * if it is good, clean up the adjustment, perform correctness testing,
+   and commit/PR it to your program's revision control; if not, try
+   something else
+ * repeat
+
+Configuration Matters
+---------------------
+
+The first thing to check is that you are using Chapel in its
+highest-performing configuration for your system.
+
+ * use ``--fast`` when compiling Chapel programs for performance
+
+ * use the default configuration rather than the quickstart configuration
+
+   * In particular, ``CHPL_TASKS=qthreads`` is generally faster than
+     ``CHPL_TASKS=fifo``, and ``CHPL_TARGET_MEM=jemalloc`` is generally
+     faster than ``CHPL_TARGET_MEM=cstdlib``
+
+   * ``CHPL_TARGET_COMPILER=llvm`` might or might not be faster than
+     using the C backend with something like
+     ``CHPL_TARGET_COMPILER=gnu``; however the performance of the LLVM
+     backend is more reliable
+
+     * if you are using a system LLVM, it's a good idea to match the
```

Comment:
Given the voice / apparenty audience of this guide, I think "system LLVM" might be confusing. Consider something like "system LLVM (as opposed to the version of LLVM bundled with Chapel)". later, you can just use "bundled" to refer to the version in the release.